lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,939 @@
+	// will stop the receive goroutine.
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+
+		for {
+			req, err := r.stream.Recv()
+			if err != nil {
+				wrappedErr := fmt.Errorf("error receiving "+
+					"request from watch-only node: %v", err)
+
+				// Send the error to the error channel, given
+				// that we're still listening on the channel.
+				select {
+				case errChan <- wrappedErr:
+				case <-streamCtx.Done():
+				case <-r.quit:
+				}
+
+				return
+			}
+
+			select {
+			case <-streamCtx.Done():
+				return
+
+			case <-r.quit:
+				return
+
+			case reqChan <- req:
+			}
+		}
+	}()
+
+	for {
+		log.Tracef("Waiting for a request from the watch-only node")
+
+		select {
+		case req := <-reqChan:
+			// Process the received request.
+			err := r.handleRequest(streamCtx, req)
+			if err != nil {
+				return err
+			}
+
+		case <-r.quit:
+			return ErrShuttingDown
+
+		case <-streamCtx.Done():
+			return streamCtx.Err()
+
+		case err := <-errChan:
+			return err
+		}
+	}
+}
+
+// handleRequest processes the received request from the watch-only node, and
+// sends the corresponding response back.
+func (r *RemoteSignerClient) handleRequest(streamCtx context.Context,
+	req *walletrpc.SignCoordinatorRequest) error {
+
+	log.Debugf("Processing a request from watch-only node of type: %T",
+		req.GetSignRequestType())
+
+	// Process the request.
+	resp, err := r.process(streamCtx, req)
+	if err != nil {
+		errStr := "error processing the request in the remote " +
+			"signer: " + err.Error()
+
+		log.Errorf(errStr)
+
+		// If we fail to process the request, we will send a SignerError
+		// back to the watch-only node, indicating the nature of the
+		// error.
+		eType := &walletrpc.SignCoordinatorResponse_SignerError{
+			SignerError: &walletrpc.SignerError{
+				Error: errStr,
+			},
+		}
+
+		resp = &walletrpc.SignCoordinatorResponse{
+			RefRequestId:     req.GetRequestId(),
+			SignResponseType: eType,
+		}
+	}
+
+	// Send the response back to the watch-only node.
+	err = r.sendResponse(streamCtx, resp)
+	if err != nil {
+		return fmt.Errorf("error sending response to watch-only "+
+			"node: %w", err)
+	}
+
+	log.Tracef("Sent the following response to watch-only node: %v", resp)
+
+	return nil
+}
+
+// process sends the passed request on to the appropriate server for processing
+// it, and returns the response.
+func (r *RemoteSignerClient) process(ctx context.Context,
+	req *walletrpc.SignCoordinatorRequest) (
+	*walletrpc.SignCoordinatorResponse, error) {
+
+	requestID := req.GetRequestId()
+
+	//nolint:lll
+	switch reqType := req.GetSignRequestType().(type) {
+	case *walletrpc.SignCoordinatorRequest_SharedKeyRequest:
+		resp, err := r.signerServer.DeriveSharedKey(
+			ctx, reqType.SharedKeyRequest,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		rType := &walletrpc.SignCoordinatorResponse_SharedKeyResponse{
+			SharedKeyResponse: resp,
+		}
+
+		signResp := &walletrpc.SignCoordinatorResponse{
```

Comment:
We could always create this message at the top, then just set `signResp.SignResponseType = rType`.